---
title: "Functional enrichment"
output: html_notebook
---
##libraries
```{r}
source("https://bioconductor.org/biocLite.R")
biocLite("rtracklayer")
library(tidyverse)
library(reutils)
```

```{r}
`%ni%` = Negate(`%in%`)

#this function converts genomes annotation files (.gff) from NCBI to a readable table
gff.parse <- function(x) {
  ids <- list()
  x %>% filter(type == "gene" | type == "CDS" | type == "tRNA" | type == "rRNA" ) -> x
  
  if ("old_locus_tag" %in% colnames(x)) {
    for (i in 1:(nrow(x)-1)) {
      if (all(x[i,2:4] == x[i+1,2:4])) {
      vec <- data.frame("chr"=x[["seqnames"]][i],
                        "acc"= x[["protein_id"]][(i+1)], 
                        "locus_tag" = x[["locus_tag"]][i], 
                        "old_locus_tag" = x[["old_locus_tag"]][i], 
                        "length_nt"= x[["width"]][i], 
                        "start"= x[["start"]][i], 
                        "end"= x[["end"]][i],
                        "strand" = x[["strand"]][i],
                        "type"= x[["gene_biotype"]][i],
                        "annotation"= x[["product"]][(i+1)],
                        stringsAsFactors = F)
      ids[[i]] <- vec
      }}} 
  else if ("gene" %in% colnames(x)) {
    for (i in 1:(nrow(x)-1)) {
      if (all(x[i,2:4] == x[i+1,2:4])) {
      vec <- data.frame("chr"=x[["seqnames"]][i],
                        "acc"= x[["protein_id"]][(i+1)], 
                        "locus_tag" = x[["locus_tag"]][i], 
                        "gene_name" = x[["gene"]][i], 
                        "length_nt"= x[["width"]][i], 
                        "start"= x[["start"]][i], 
                        "end"= x[["end"]][i],
                        "strand" = x[["strand"]][i],
                        "type"= x[["gene_biotype"]][i],
                        "annotation"= x[["product"]][(i+1)],
                        stringsAsFactors = F)
      ids[[i]] <- vec
      }}}
  
  df <- bind_rows(ids)
  if (length(na.omit(unique(x[["protein_id"]]))) == length(na.omit(unique(df[["acc"]])))) {
    print("all protein ids accounted for")
  }
  return(df)
}

getNOG <- function(x, colname = "eggNOGs", sep = " ", factors = FALSE) {
  lis <- list()
  for (i in 1:length(x[[colname]])) {
    vec <- c(capture.output(cat(unlist(str_extract_all(x[[colname]][i], "[a-zA-Z0-9]*@NOG")), sep = sep)))
    lis[[i]] <- vec
    column.names <- "NOGs"
    } 
  df <- as.data.frame(do.call(rbind, lis), stringsAsFactors = FALSE)
  colnames(df) <- column.names
  df <- cbind("query" = x[["query"]], df, "COG_category" = x[["COG Cat."]], "HMM_description" = x[["eggNOG HMM Desc."]])
  return(df)
}

#this function gets sequences based on your acc list
getseqs <- function(x, filename) {
  #x must be a vector of acc or gi numbers.
  require(reutils)
  path <- file.path(getwd(), filename)
  uid1 <- esearch(x, "protein", rettype = "uilist", usehistory = T)
  efetch(uid1, "protein", "fasta", "text", outfile = filename)
  print(paste("sequences can be found at ", path))
}
#TESTS FOR FUNCTIONAL ENRICHMENT
#performs hypergeometric test on provided subset of genes/proteins relative to the genome
#code adapted from Keely Dulmage
nogtest <- function(namelist,nogfile,pvalue, cutoff = 5) {
  #namelist is a vector of protein on gene names you wnat to test for enrichment
  #nogfile is the genome-wide GETNOG output
  #p-value is significance threshold desired
  #cutoff preset prevents functional categories with less than the designated number of genes/proteins being displayed 
  
  nogs <- nogfile[nogfile[["query"]] %in% namelist,]
  clust <-  table(nogs[["COG_category"]])
  resm <- matrix(0,length(clust),3) #create 0 matrix
  res <- data.frame(resm)  #make 0 df
  rownames(res) <- names(clust)
  colnames(res) <- c("probability", "expected","count")
  all <- table(nogfile[["COG_category"]][nogfile[["COG_category"]] %in% nogs[["COG_category"]]])
  tot <- sum(table(nogfile[["COG_category"]]))
  #print(tot); print(all); print(clust)
  for (i in 1:length(clust)){   #calc expected frequencies and pval by hypergeo and append to DF
    
    res[i,1] <- signif(phyper(clust[i], all[i], tot-all[i], nrow(nogs),lower.tail=F), digits = 4)
    res[i,2] <- signif(all[i]*(nrow(nogs)/tot), digits = 4)
    res[i,3] <- clust[i]
  }
  fin <- subset(res, probability <= pvalue & count >= cutoff)
  fin$COG_category <- rownames(fin)
  fin <- fin[, c("COG_category", "probability", "expected", "count")]
  return(fin)
}

#Use the following function to look at the genes in your cluster associated with a particular COG.
nogset= function(namelist,nogfile, cog.category) {
  subset(nogfile, is.element(nogfile$query, namelist) & is.element(nogfile$COG_category, cog.category)==TRUE)
}
```

##Getting fasta sequences from gene names/locus tags
```{r}
#read in your .gff file and list of significant genes here
my.genes <- read_csv("H26_sig.csv")
my.gff <- as.data.frame(rtracklayer::import.gff("DS2_genome.gff"))

#execute function
gff.table <- gff.parse(my.gff)

#convert from locus tags to GI accession numbers
my.accs <- gff.table[gff.table$old_locus_tag %in% my.genes$gene,]$acc

#check length
length(my.accs) == length(my.genes$gene)
my.genes[my.genes$gene %ni% gff.table$old_locus_tag,]

#get sequences
getseqs(my.accs, paste(Sys.Date(), "_seqs.for.eggnog.faa", sep = ""))
```

These differentially expressed genes do not exist in the gff table
```{r}
my.genes$gene[my.genes$gene %ni% gff.table$old_locus_tag]
```

##Parsing the eggNOG output file
```{r}
my.eggNOG <- read_csv("emapper.csv")
my.genome.NOGs <- getNOG(my.eggNOG)
```

these genes were identified as differentially expressed, but are not present in the eggNOG database
```{r}
gff.table[gff.table$acc %in% my.accs[my.accs %ni% my.genome.NOGs$query],]
```


##Functional enrichment
```{r}
#append unknown function annotation to proteins with missing COG classifications
my.genome.NOGs$COG_category[is.na(my.genome.NOGs$COG_category)] <- "S"
my.genome.NOGs$HMM_description[is.na(my.genome.NOGs$HMM_description)] <- "function unknown, manually assigned by RKH"

#calculate enrichment of all functional categories in the DEGs
my.genes.hypg <-  nogtest(my.accs, my.genome.NOGs, 1, cutoff = 1)

#correct for multiple testing, using FDR. 
my.genes.hypg$p.adj <- p.adjust(my.genes.hypg$probability, method = "fdr")

my.genes.hypg
```

```{r}
#get proteins with specific function, where "X" is the single letter functional code

ls <- list()
counter <- 1
for (i in my.genes.hypg$COG_category){
  ls[[counter]] <- nogset(my.accs, my.genome.NOGs, i)
  counter <- counter +1
}
my.sig.proteins <- bind_rows(ls)

p.vals <- vector()
for (i in 1:length(my.sig.proteins$COG_category)) {
  cog <- my.sig.proteins$COG_category[i]
  p.vals[i] <- filter(my.genes.hypg, my.genes.hypg$COG_category == cog)$p.adj
}

#convert those proteins from acc back to locus_tags
my.sig.proteins$old_locus_tag <- gff.table$old_locus_tag[gff.table$acc %in% my.sig.proteins$query]
my.sig.proteins$COG.padj <- p.vals

(my.sig.proteins<- my.sig.proteins[c(5,1,3,6,4,2)])

#write out files
write_csv(my.sig.proteins, paste(Sys.Date(), "sig_genes_enriched_fxn.csv", sep = ""))
```
